pallets/timestamp/src/lib.rs

```diff
@@ -166,16 +166,21 @@ decl_module! {
             T::WeightInfo::set(),
             DispatchClass::Mandatory
         )]
-        fn set(origin, #[compact] now: T::Moment) {
+        fn set(origin, #[compact] timestamp: T::Moment) {
             ensure_none(origin)?;
             assert!(!<Self as Store>::DidUpdate::exists(), "Timestamp must be updated only once in the block");
-            //let prev = Self::now();
-            //FIXME bring back the check
-            // assert!(
-            // 	prev.is_zero() || now >= prev + T::MinimumPeriod::get(),
-            // 	"Timestamp must increment by at least <MinimumPeriod> between sequential blocks"
-            // );
-            <Self as Store>::Now::put(now);
+
+            // in Mangata extrinsics stored in block N are executed in block N+1
+            let now = Self::now();
+            let block_time = T::MinimumPeriod::get();
+            let allowed_period_begin = now + block_time + block_time;
+
+            assert!(
+            	now.is_zero() || timestamp >= allowed_period_begin,
```

Comment:
At the time of execution of `set` method, `timestamp` is 6 seconds ago, and `now` is 12 seconds ago, and with current calculations `allowed_period_begin` will be at the present moment (since -12 + 6 + 6). Thus, `timestamp` will be smaller than `allowed_period_begin`.
What am I missing here?

Reply:
i just checked and it seems like the problem why assertion was failing was not due to increased delay but to the fact that for the second block `now - prev == 0`. 
```
May 08 10:45:54.041  INFO apply_extrinsic: prev:0 now:1620463554004 diff:1620463554004 minimum_period:3000    {ext}
May 08 10:46:00.007  INFO apply_extrinsic: prev:1620463554004 now:1620463554004 diff:0 minimum_period:3000    {ext}
May 08 10:46:06.007  INFO apply_extrinsic: prev:1620463554004 now:1620463560000 diff:5996 minimum_period:3000    {ext}
May 08 10:46:12.007  INFO apply_extrinsic: prev:1620463560000 now:1620463566000 diff:6000 minimum_period:3000    {ext}
May 08 10:46:18.008  INFO apply_extrinsic: prev:1620463566000 now:1620463572000 diff:6000 minimum_period:3000    {ext}
May 08 10:46:24.007  INFO apply_extrinsic: prev:1620463572000 now:1620463578000 diff:6000 minimum_period:3000    {ext}
May 08 10:46:30.007  INFO apply_extrinsic: prev:1620463578000 now:1620463584000 diff:6000 minimum_period:3000    {ext}
May 08 10:46:36.007  INFO apply_extrinsic: prev:1620463584000 now:1620463590000 diff:6000 minimum_period:3000    {ext}
```

after adding `(now-prev) == 0` both node and tests are working fine